lib/libdevctl/devctl.3
.\"
.\" Copyright (c) 2014 John Baldwin <jhb@FreeBSD.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: head/lib/libdevctl/devctl.3 298709 2016-04-27 16:33:17Z jhb $
.\"
.Dd April 27, 2016
.Dt DEVCTL 3
.Os
.Sh NAME
.Nm devctl ,
.Nm devctl_attach ,
.Nm devctl_delete ,
.Nm devctl_detach ,
.Nm devctl_disable ,
.Nm devctl_enable ,
.Nm devctl_rescan ,
.Nm devctl_resume ,
.Nm devctl_set_driver ,
.Nm devctl_suspend
.Nd device control library
.Sh LIBRARY
.Lb libdevctl
.Sh SYNOPSIS
.In devctl.h
.Ft int
.Fn devctl_attach "const char *device"
.Ft int
.Fn devctl_delete "const char *device" "bool force"
.Ft int
.Fn devctl_detach "const char *device" "bool force"
.Ft int
.Fn devctl_disable "const char *device" "bool force_detach"
.Ft int
.Fn devctl_enable "const char *device"
.Ft int
.Fn devctl_rescan "const char *device"
.Ft int
.Fn devctl_resume "const char *device"
.Ft int
.Fn devctl_set_driver "const char *device" "const char *driver" "bool force"
.Ft int
.Fn devctl_suspend "const char *device"
.Sh DESCRIPTION
The
.Nm
library adjusts the state of devices in the kernel's internal device
hierarchy.
Each control operation accepts a
.Fa device
argument that identifies the device to adjust.
The
.Fa device
may be specified as either the name of an existing device or as a
bus-specific address.
The following bus-specific address formats are currently supported:
.Bl -tag -offset indent
.It Sy pci Ns Fa domain Ns : Ns Fa bus Ns : Ns Fa slot Ns : Ns Fa function
A PCI device with the specified
.Fa domain ,
.Fa bus ,
.Fa slot ,
and
.Fa function .
.It Sy pci Ns Fa bus Ns : Ns Fa slot Ns : Ns Fa function
A PCI device in domain zero with the specified
.Fa bus ,
.Fa slot ,
and
.Fa function .
.It Fa handle
A device with an ACPI handle of
.Fa handle .
The handle must be specified as an absolute path and must begin with a
.Dq \e .
.El
.Pp
The
.Fn devctl_attach
function probes a device and attaches a suitable device driver if one is
found.
.Pp
The
.Fn devctl_detach
function detaches a device from its current device driver.
The device is left detached until either a new driver for its parent
bus is loaded or the device is explicitly probed via
.Fn devctl_attach .
If
.Fa force
is true,
the current device driver will be detached even if the device is busy.
.Pp
The
.Fn devctl_delete
function deletes a device from the device tree.
No
If
.Fa force
is true,
the device is deleted even if the device is physically present.
.Pp
The
.Fn devctl_disable
function disables a device.
If the device is currently attached to a device driver,
the device driver will be detached from the device,
but the device will retain its current name.
If
.Fa force_detach
is true,
the current device driver will be detached even if the device is busy.
The device will remain disabled and detached until it is explicitly enabled
via
.Fn devctl_enable .
.Pp
The
.Fn devctl_enable
function re-enables a disabled device.
The device will probe and attach if a suitable device driver is found.
.Pp
The
.Fn devctl_suspend
function suspends a device.
This may include placing the device in a reduced power state,
but any device driver currently attached to the device will remain attached.
.Pp
The
.Fn devctl_resume
function resumes a suspended device to a fully working state.
.Pp
The
.Fn devctl_set_driver
function attaches a device driver named
.Fa driver
to a device.
If the device is already attached and
.Fa force
is false,
the request will fail.
If the device is already attached and
.Fa force
is true,
the device will be detached from its current device driver before it is
attached to the new device driver.
.Pp
The
.Fn devctl_rescan
function rescans a bus device checking for devices that have been added or
removed.
.Sh RETURN VALUES
.Rv -std devctl_attach devctl_delete devctl_detach devctl_disable \
devctl_enable devctl_suspend devctl_rescan devctl_resume devctl_set_driver
.Sh ERRORS
In addition to specific errors noted below,
all of the
.Nm
functions may fail for any of the errors described in
.Xr open 2
as well as:
.Bl -tag -width Er
.It Bq Er EINVAL
The device name is too long.
.It Bq Er ENOENT
No existing device matches the specified name or location.
.It Bq Er EPERM
The current process is not permitted to adjust the state of
.Fa device .
.El
.Pp
The
.Fn devctl_attach
function may fail if:
.Bl -tag -width Er
.It Bq Er EBUSY
The device is already attached.
.It Bq Er ENOMEM
An internal memory allocation request failed.
.It Bq Er ENXIO
The device is disabled.
.It Bq Er ENXIO
No suitable driver for the device could be found,
or the driver failed to attach.
.El
.Pp
The
.Fn devctl_detach
function may fail if:
.Bl -tag -width Er
.It Bq Er EBUSY
The current device driver for
.Fa device
is busy and cannot detach at this time.
Note that some drivers may return this even if
.Fa force
is true.
.It Bq Er ENXIO
The device is not attached to a driver.
.It Bq Er ENXIO
The current device driver for
.Fa device
does not support detaching.
.El
.Pp
The
.Fn devctl_enable
function may fail if:
.Bl -tag -width Er
.It Bq Er EBUSY
The device is already enabled.
.It Bq Er ENOMEM
An internal memory allocation request failed.
.It Bq Er ENXIO
No suitable driver for the device could be found,
or the driver failed to attach.
.El
.Pp
The
.Fn devctl_disable
function may fail if:
.Bl -tag -width Er
.It Bq Er EBUSY
The current device driver for
.Fa device
is busy and cannot detach at this time.
Note that some drivers may return this even if
.Fa force_detach
is true.
.It Bq Er ENXIO
The device is already disabled.
.It Bq Er ENXIO
The current device driver for
.Fa device
does not support detaching.
.El
.Pp
The
.Fn devctl_suspend
function may fail if:
.Bl -tag -width Er
.It Bq Er EBUSY
The device is already suspended.
.It Bq Er EINVAL
The device to be suspended is the root bus device.
.El
.Pp
The
.Fn devctl_resume
function may fail if:
.Bl -tag -width Er
.It Bq Er EINVAL
The device is not suspended.
.It Bq Er EINVAL
The device to be resumed is the root bus device.
.El
.Pp
The
.Fn devctl_set_driver
function may fail if:
.Bl -tag -width Er
.It Bq Er EBUSY
The device is currently attached to a device driver and
.Fa force
is false.
.It Bq Er EBUSY
The current device driver for
.Fa device
is busy and cannot detach at this time.
.It Bq Er EFAULT
The
.Fa driver
argument points outside the process' allocated address space.
.It Bq Er ENOENT
No device driver with the requested name exists.
.It Bq Er ENOMEM
An internal memory allocation request failed.
.It Bq Er ENXIO
The device is disabled.
.It Bq Er ENXIO
The new device driver failed to attach.
.El
.Pp
The
.Fn devctl_rescan
function may fail if:
.Bl -tag -width Er
.It Bq Er ENXIO
The device is not attached to a driver.
.It Bq Er ENXIO
The bus driver does not support rescanning.
.El
.Pp
The
.Fn devctl_delete
function may fail if:
.Bl -tag -width Er
.It Bq Er EBUSY
The device is physically present and
.Fa force
is false.
.It Bq Er EINVAL
.Fa dev
is the root device of the device tree.
.El
.Sh SEE ALSO
.Xr devinfo 3 ,
.Xr devstat 3 ,
.Xr devctl 8
.Sh HISTORY
The
.Nm
library first appeared in
.Fx 10.3 .
.Sh BUGS
If a device is suspended individually via
.Fn devctl_suspend
and the entire machine is subsequently suspended,
the device will be resumed when the machine resumes.


usr.sbin/devctl/devctl.c
/*-
 * Copyright (c) 2014 John Baldwin <jhb@FreeBSD.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: head/usr.sbin/devctl/devctl.c 298709 2016-04-27 16:33:17Z jhb $");

#include <sys/linker_set.h>
#include <devctl.h>
#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <unistd.h>

struct devctl_command {
	const char *name;
	int (*handler)(int ac, char **av);
};

#define	DEVCTL_DATASET(name)	devctl_ ## name ## _table

#define	DEVCTL_COMMAND(set, name, function)				\
	static struct devctl_command function ## _devctl_command =	\
	{ #name, function };						\
	DATA_SET(DEVCTL_DATASET(set), function ## _devctl_command)

#define	DEVCTL_TABLE(set, name)						\
	SET_DECLARE(DEVCTL_DATASET(name), struct devctl_command);	\
									\
	static int							\
	devctl_ ## name ## _table_handler(int ac, char **av)		\
	{								\
		return (devctl_table_handler(SET_BEGIN(DEVCTL_DATASET(name)), \
		    SET_LIMIT(DEVCTL_DATASET(name)), ac, av));		\
	}								\
	DEVCTL_COMMAND(set, name, devctl_ ## name ## _table_handler)

static int	devctl_table_handler(struct devctl_command **start,
    struct devctl_command **end, int ac, char **av);

SET_DECLARE(DEVCTL_DATASET(top), struct devctl_command);

DEVCTL_TABLE(top, set);

static void
usage(void)
{
	fprintf(stderr, "%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n",
	    "usage: devctl attach device",
	    "       devctl detach [-f] device",
	    "       devctl disable [-f] device",
	    "       devctl enable device",
	    "       devctl suspend device",
	    "       devctl resume device",
	    "       devctl set driver [-f] device driver",
	    "       devctl rescan device",
	    "       devctl delete [-f] device");
	exit(1);
}

static int
devctl_table_handler(struct devctl_command **start,
    struct devctl_command **end, int ac, char **av)
{
	struct devctl_command **cmd;

	if (ac < 2) {
		warnx("The %s command requires a sub-command.", av[0]);
		return (EINVAL);
	}
	for (cmd = start; cmd < end; cmd++) {
		if (strcmp((*cmd)->name, av[1]) == 0)
			return ((*cmd)->handler(ac - 1, av + 1));
	}

	warnx("%s is not a valid sub-command of %s.", av[1], av[0]);
	return (ENOENT);
}

static int
help(int ac __unused, char **av __unused)
{

	usage();
	return (0);
}
DEVCTL_COMMAND(top, help, help);

static int
attach(int ac, char **av)
{

	if (ac != 2)
		usage();
	if (devctl_attach(av[1]) < 0)
		err(1, "Failed to attach %s", av[1]);
	return (0);
}
DEVCTL_COMMAND(top, attach, attach);

static void
detach_usage(void)
{

	fprintf(stderr, "usage: devctl detach [-f] device\n");
	exit(1);
}

static int
detach(int ac, char **av)
{
	bool force;
	int ch;

	force = false;
	while ((ch = getopt(ac, av, "f")) != -1)
		switch (ch) {
		case 'f':
			force = true;
			break;
		default:
			detach_usage();
		}
	ac -= optind;
	av += optind;

	if (ac != 1)
		detach_usage();
	if (devctl_detach(av[0], force) < 0)
		err(1, "Failed to detach %s", av[0]);
	return (0);
}
DEVCTL_COMMAND(top, detach, detach);

static void
disable_usage(void)
{

	fprintf(stderr, "usage: devctl disable [-f] device\n");
	exit(1);
}

static int
disable(int ac, char **av)
{
	bool force;
	int ch;

	force = false;
	while ((ch = getopt(ac, av, "f")) != -1)
		switch (ch) {
		case 'f':
			force = true;
			break;
		default:
			disable_usage();
		}
	ac -= optind;
	av += optind;

	if (ac != 1)
		disable_usage();
	if (devctl_disable(av[0], force) < 0)
		err(1, "Failed to disable %s", av[0]);
	return (0);
}
DEVCTL_COMMAND(top, disable, disable);

static int
enable(int ac, char **av)
{

	if (ac != 2)
		usage();
	if (devctl_enable(av[1]) < 0)
		err(1, "Failed to enable %s", av[1]);
	return (0);
}
DEVCTL_COMMAND(top, enable, enable);

static int
suspend(int ac, char **av)
{

	if (ac != 2)
		usage();
	if (devctl_suspend(av[1]) < 0)
		err(1, "Failed to suspend %s", av[1]);
	return (0);
}
DEVCTL_COMMAND(top, suspend, suspend);

static int
resume(int ac, char **av)
{

	if (ac != 2)
		usage();
	if (devctl_resume(av[1]) < 0)
		err(1, "Failed to resume %s", av[1]);
	return (0);
}
DEVCTL_COMMAND(top, resume, resume);

static void
set_driver_usage(void)
{

	fprintf(stderr, "usage: devctl set driver [-f] device driver\n");
	exit(1);
}

static int
set_driver(int ac, char **av)
{
	bool force;
	int ch;

	force = false;
	while ((ch = getopt(ac, av, "f")) != -1)
		switch (ch) {
		case 'f':
			force = true;
			break;
		default:
			set_driver_usage();
		}
	ac -= optind;
	av += optind;

	if (ac != 2)
		set_driver_usage();
	if (devctl_set_driver(av[0], av[1], force) < 0)
		err(1, "Failed to set %s driver to %s", av[0], av[1]);
	return (0);
}
DEVCTL_COMMAND(set, driver, set_driver);

static int
rescan(int ac, char **av)
{

	if (ac != 2)
		usage();
	if (devctl_rescan(av[1]) < 0)
		err(1, "Failed to rescan %s", av[1]);
	return (0);
}
DEVCTL_COMMAND(top, rescan, rescan);

static void
delete_usage(void)
{

	fprintf(stderr, "usage: devctl delete [-f] device\n");
	exit(1);
}

static int
delete(int ac, char **av)
{
	bool force;
	int ch;

	force = false;
	while ((ch = getopt(ac, av, "f")) != -1)
		switch (ch) {
		case 'f':
			force = true;
			break;
		default:
			delete_usage();
		}
	ac -= optind;
	av += optind;

	if (ac != 1)
		delete_usage();
	if (devctl_delete(av[0], force) < 0)
		err(1, "Failed to delete %s", av[0]);
	return (0);
}
DEVCTL_COMMAND(top, delete, delete);

int
main(int ac, char *av[])
{
	struct devctl_command **cmd;

	if (ac == 1)
		usage();
	ac--;
	av++;

	SET_FOREACH(cmd, DEVCTL_DATASET(top)) {
		if (strcmp((*cmd)->name, av[0]) == 0) {
			if ((*cmd)->handler(ac, av) != 0)
				return (1);
			else
				return (0);
		}
	}
	warnx("Unknown command %s.", av[0]);
	return (1);
}
